text/3724-theupdateframework.md

```diff
@@ -0,0 +1,341 @@
+
+## Repository Workflows
+[tuf-on-ci](https://github.com/theupdateframework/tuf-on-ci) shall be used for workflows on each repository.
+
+tuf-on-ci is a set of CI tools which are integrated with GitHub Actions for providing TUF quorum and hardware key support for managing TUF repositories via pull requests on GitHub. This is a ready made and production ready suite of CI that is used by the sigstore root signing for managing TUF quorums.
+
+### (tuf-root) Rustup Release
+
+- A signing event will be triggered via pull request in the `tuf-root` repository, which is assigned to the `Rustup Role` quorum, who must sign the request.
+
+### (tuf-root) Stable/Beta Release
+
+- A signing event will be triggered via pull request in the `tuf-root` repository, which is assigned to the `Release Role` quorum, who must sign the request.
+
+### (tuf-root) Nightly Release
+
+- An online signing delegate key will be utilized for signing updates to the repository. This key shall live in AWS KMS and allow for automated signing within the `tuf-root` repository.
+
+### (tuf-crates) Crate Release/Yank
+
+- An online signing delegate key will be utilized for signing updates to the repository. This key shall live in AWS KMS and allow for automated signing within the `tuf-crates` repository.
+
+### (tuf-root + tuf-crates) Crates.io Membership Change
+
+- The crates.io team will update the root role in the `tuf-crates` repository, triggering a signing event that the existing crates.io team must sign via Pull Request.
+- An update to the tuf-crates-root.json file will occur in the `tuf-root` repository, which shall trigger a new singing event Pull Request, which the root quorum must perform.
+
+## (cargo-tuf-lib) Standard TUF Implementation
+
+We propose creating a new crate, `cargo-tuf-lib`, which shall be used by both Cargo and Rustup for doing TUF synchronization and update procedures. This library shall be a shim wrapper around the `rust-tuf` crate (https://github.com/rustfoundation/rust-tuf), providing a simplified and shared interface for doing synchronization and verification of the TUF repositories and their files. 
+
+The API surface of this crate is to be determined upon implementation in Cargo and Rustup. However, because both tools will need to perform synchronization and validation against the tuf-root repository, they shall used this shared interface to guarantee compatibility.
+
+This API will include operations to sync the TUF repositories efficiently, and to perform a verified download of an object.
+
+## TAP-16 Implementation
+
+We're proposing to use [TAP-16](https://github.com/theupdateframework/taps/blob/master/tap16.md) to provide efficient update checking and download sizes. TAP-16 uses Merkle trees rather than full lists for the download of a snapshot of the inventory of a repository (`snapshot.json`). We want to ensure that, as crates.io grows, the total size clients have to download when checking for updates remains small.
```

Comment:
In practice, what is the expected impact for this for crates.io's current size and if it grew to the size of some of the larger registries for other languages?

Reply:
Yeah it would be good to understand the time and space complexities of the operations involved.

Does TAP-16 prevent a full snapshot being necessary at all, or does it just reduce the download size? I think it should be possible to validate a set of dependencies without having to have a full snapshot of the repository...

Reply:
I completely agree that these are the critical questions. Based on my reading of TAP-16, the answers entirely depend on the implementation details that are specified in the [POUF](https://github.com/theupdateframework/taps/blob/master/tap11.md). I think we need these answers in order to stabilize this functionality.